File: figures/05-dynamic-programming/snippet-fib-bottom-up.c
static bool fib_bottom_up(int n, long long *out) {
    if (!out || n < 0 || n > 92) return false;
    if (n <= 1) { *out = n; return true; }
    long long prev2 = 0, prev1 = 1;
    for (int i = 2; i <= n; ++i) {
        long long current = prev2 + prev1;
        prev2 = prev1; prev1 = current;
    }
    *out = prev1; return true;
}

File: figures/05-dynamic-programming/snippet-fib-bottom-up.py
def fib_bottom_up(n):
    if n <= 1:
        return n
    prev2, prev1 = 0, 1
    for _ in range(2, n + 1):
        current = prev2 + prev1
        prev2, prev1 = prev1, current
    return prev1
